streaming/text/convert/laion/laion400m/convert_and_upload.py

```diff
@@ -0,0 +1,276 @@
+        return 0
+    else:
+        return int(x)
+
+
+def get_float(x: float) -> float:
+    """Get a float field from pandas.
+
+    Args:
+        x (float): The pandas field.
+
+    Returns:
+        float: The normalized field.
+    """
+    return x
+
+
+def get_bytes(x: Optional[bytes]) -> bytes:
+    """Get a bytes field from pandas.
+
+    Args:
+        x (bytes, optional): The pandas field.
+
+    Returns:
+        float: The normalized field.
+    """
+    return x or b''
+
+
+def get_str(x: Optional[str]) -> str:
+    """Get a str field from pandas.
+
+    Args:
+        x (str, optional): The pandas field.
+
+    Returns:
+        str: The normalized field.
+    """
+    return x or ''
+
+
+def convert(parquet_filename: str, mds_dirname: str, hashes: List[str]) -> None:
+    """Convert a parquet shard to MDS shard with an index.
+
+    Args:
+        parquet_filename (str): Filename of the input parquet shard.
+        mds_dirname (str): Dirname containing the output MDS shard and index.
+        hashes (List[str]): List of hashes to apply to the MDS shard, if any.
+    """
+    columns = {
+        'nsfw': 'str',
+        'similarity': 'float64',
+        'license': 'str',
+        'caption': 'str',
+        'url': 'str',
+        'key': 'str',
+        'status': 'str',
+        'error_message': 'str',
+        'width': 'int32',
+        'height': 'int32',
+        'original_width': 'int32',
+        'original_height': 'int32',
+        'exif': 'str',
+        'jpg': 'bytes',
+    }
+    compression = None  # Don't compress because the vast majority of the data is JPEG.
+    size_limit = None  # Put everything in one shard (1:1 mapping of parquet to MDS shards).
+
+    with MDSWriter(mds_dirname, columns, compression, hashes, size_limit) as out:
+        table = pq.read_table(parquet_filename)
+        nr = table.num_rows
+        table = table.to_pandas()
+        for i in range(nr):
+            x = table.iloc[i]
+            sample = {
+                'nsfw': get_str(x['NSFW']),
+                'similarity': get_float(x['similarity']),
+                'license': get_str(x['LICENSE']),
+                'caption': get_str(x['caption']),
+                'url': get_str(x['url']),
+                'key': get_str(x['key']),
+                'status': get_str(x['status']),
+                'error_message': get_str(x['error_message']),
+                'width': get_int(x['width']),
+                'height': get_int(x['height']),
+                'original_width': get_int(x['original_width']),
+                'original_height': get_int(x['original_height']),
+                'exif': get_str(x['exif']),
+                'jpg': get_bytes(x['jpg']),
+            }
+            out.write(sample)
+
+
+def upload(local: str, remote: str) -> None:
+    """Upload a shard to remote storage.
+
+    Args:
+        local (str): Path on local filesystem.
+        remote (str): Path on remote filesystem.
+    """
+    assert ' ' not in local
```

Comment:
Please avoid all the assert statements. Thanks!

Reply:
done